lib/Migration/Version1140Date20220630113356.php

```diff
@@ -0,0 +1,61 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @copyright Copyright (c) 2022 Your name <your@email.com>
+ *
+ * @author Your name <your@email.com>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Mail\Migration;
+
+use Closure;
+use OCP\DB\ISchemaWrapper;
+use OCP\Migration\IOutput;
+use OCP\Migration\SimpleMigrationStep;
+
+class Version1140Date20220630113356 extends SimpleMigrationStep {
+
+	/**
+	 * @param IOutput $output
+	 * @param Closure $schemaClosure The `\Closure` returns a `ISchemaWrapper`
+	 * @param array $options
+	 * @return null|ISchemaWrapper
+	 */
+	public function changeSchema(IOutput $output, Closure $schemaClosure, array $options): ?ISchemaWrapper {
+		/** @var ISchemaWrapper $schema */
+		$schema = $schemaClosure();
+
+		$messagesTable = $schema->getTable('mail_messages');
+		$messagesTable->addColumn('imip_message', 'boolean', [
+			'notnull' => false,
+			'default' => false,
+		]);
+		$messagesTable->addColumn('imip_processed', 'boolean', [
+			'notnull' => false,
+			'default' => false,
+		]);
+		$messagesTable->addColumn('imip_error', 'boolean', [
+			'notnull' => false,
+			'default' => false,
+		]);
```

Comment:
something to be written down somewhere: write migrations in a way they can be run more than once.

Imagine adding the second column fails. Then oc_migrations *won't* have this migration and nextcloud will try to run it another time. Then you get an error because `imip_message` already exists. That is because DDL are not transactional.

So write it like
```
if (!$table->hasColumn(xyz) {
  $table->addColumn(...);
}
```

Reply:
Good point, I will add this!